src/github.com/getlantern/enproxy/conn_test.go

```diff
@@ -62,18 +60,80 @@ func TestBadBuffered(t *testing.T) {
 	doTestBad(true, t)
 }
 
+func TestIdle(t *testing.T) {
+	idleTimeout := 100 * time.Millisecond
+
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
+	_, err = Dial(httpAddr, &Config{
+		DialProxy: func(addr string) (net.Conn, error) {
+			return net.Dial("tcp", proxyAddr)
+		},
+		NewRequest:  newRequest,
+		IdleTimeout: idleTimeout,
+	})
+	if assert.NoError(t, err, "Dialing should have succeeded") {
+		time.Sleep(idleTimeout * 2)
+		assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
+	}
+}
+
+// This test stimulates a connection leak as seen in
+// https://github.com/getlantern/lantern/issues/2174.
+func TestHTTPRedirect(t *testing.T) {
+	startProxy(t)
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			Dial: func(network, addr string) (net.Conn, error) {
+				return Dial(addr, &Config{
+					DialProxy: func(addr string) (net.Conn, error) {
+						return net.Dial("tcp", proxyAddr)
+					},
+					NewRequest: newRequest,
+				})
+			},
+			DisableKeepAlives: true,
+		},
+	}
+
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
+	resp, err := client.Head("http://www.facebook.com")
+	if assert.NoError(t, err, "Head request to facebook should have succeeded") {
+		resp.Body.Close()
+	}
+
+	assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
+}
+
 func doTestPlainText(buffered bool, t *testing.T) {
 	startServers(t)
 
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
 	conn, err := prepareConn(httpAddr, buffered, false, t)
 	if err != nil {
 		t.Fatalf("Unable to prepareConn: %s", err)
 	}
-	defer conn.Close()
+	defer func() {
+		err := conn.Close()
+		assert.Nil(t, err, "Closing conn should succeed")
+		assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
```

Comment:
Or maybe we only want to match a CLOSE_WAIT if it is matched and would thus be returned?  I have no idea whether this matters, but I guess it can't hurt.  So, just put the check for (CLOSE_WAIT) inside the `if` for the other one:
```go
func (c *Counter) matchingLinesInner(out []byte) (string, int, bool) {
    lines := make([]string, 0)
    anyCloseWait := false
    for _, line := range strings.Split(string(out), "\n") {
        if strings.Contains(line, c.match) {
            if strings.HasSuffix(line, "(CLOSE_WAIT)") {
                anyCloseWait := true
            }
            lines = append(lines, line)
        }
    }
    return strings.Join(lines, "\n"), len(lines), anyCloseWait
}
```